### ISSUE TITLE
Dequeues aren't amortized O(1) when using both `unsnoc` and `uncons`

### ISSUE THREAD

**Issue by ZiyaoWei** [scalaz/scalaz#2531]
It seems `dequeue`s are not amortized O(1) when both `unsnoc` and `uncons` are called; in fact, the underlying two lists might "oscillate" such that the operations are amortized O(n). When only one of these is used it is amortized O(1).

See https://scastie.scala-lang.org/lrP3QumzQuSMQ1rQMBUGow for a demonstration.

Given this has already been there for a while it is probably enough for real world use cases - should the docs be updated?

Note that Okasaki does contain an amortized O(1) banker's deque, which we can implement if we truly want amortized constant time operations.

**Comment by TomasMikula** [comment]
Yes, I think the docs should be updated to make it clear that only consecutive dequeues from the same end are amortized _O(1)._
